activerecord/lib/active_record/store.rb

```diff
@@ -225,33 +231,45 @@ def store_accessor_for(store_attribute)
       end
 
       class HashAccessor # :nodoc:
+        def self.get(store_object, key)
+          if store_object
+            store_object[key]
+          end
+        end
+
         def self.read(object, attribute, key)
-          prepare(object, attribute)
-          object.public_send(attribute)[key]
+          store_object = object.public_send(attribute) || init(object, attribute)
+          store_object[key]
         end
 
         def self.write(object, attribute, key, value)
-          prepare(object, attribute)
-          object.public_send(attribute)[key] = value if value != read(object, attribute, key)
+          store_object = object.public_send(attribute) || init(object, attribute)
+          store_object[key] = value if value != store_object[key]
         end
 
-        def self.prepare(object, attribute)
-          object.public_send :"#{attribute}=", {} unless object.send(attribute)
+        def self.init(object, attribute)
+          store_object = {}
+          object.public_send :"#{attribute}=", store_object
+          store_object
         end
       end
 
       class StringKeyedHashAccessor < HashAccessor # :nodoc:
+        def self.get(store_object, key)
+          super store_object, Symbol === key ? key.name : key.to_s
+        end
+
         def self.read(object, attribute, key)
-          super object, attribute, key.to_s
+          super object, attribute, Symbol === key ? key.name : key.to_s
         end
 
         def self.write(object, attribute, key, value)
-          super object, attribute, key.to_s, value
+          super object, attribute, Symbol === key ? key.name : key.to_s, value
         end
       end
 
       class IndifferentHashAccessor < ActiveRecord::Store::HashAccessor # :nodoc:
-        def self.prepare(object, store_attribute)
+        def self.init(object, store_attribute)
           attribute = object.send(store_attribute)
```

Comment:
Didn't this change behavior? Previously this would have converted `attribute` from Hash to ActiveSupport::HashWithIndifferentAccess, but now this will only be called if `attribute` is nil.